File: retos-ts/25_lenguaje_magico.ts
// 5 ⭐
function execute(code: string): number {
    let acc: number = 0
    
    const processCode = (block: string) => {
        let i: number = 0;
        while (i < block.length) {
            const curr: string = block[i];

            if (curr === '+') {
                acc++;
            } else if (curr === '-') {
                acc--;
            } else if (curr === '[') {
                const end: number = block.indexOf(']', i);
                const newCode: string = block.slice(i + 1, end);

                while (acc !== 0) {
                    processCode(newCode);
                }
                i = end;
            } else if (curr === '{') {
                if (acc === 0) {
                    i = block.indexOf('}', i);
                }
            }
            i++;
        }
    };

    code = code.replaceAll('>', '');
    processCode(code);

    return acc
}

File: retos-python/25_lenguaje_magico.py
def execute(code):
    acc = 0

    def process_code(block):
        nonlocal acc
        i = 0
        while i < len(block):
            curr = block[i]

            if curr == '+':
                acc += 1
            elif curr == '-':
                acc -= 1
            elif curr == '[':
                end = block.find(']', i)
                new_code = block[i + 1:end]

                while acc != 0:
                    process_code(new_code)
                i = end
            elif curr == '{':
                if acc == 0:
                    i = block.find('}', i)
            i += 1

    code = code.replace('>', '')
    process_code(code)

    return acc